### DISCUSSION TITLE
Mayastor core isolation

### DISCUSSION THREAD

**Discussion by matthiasarndt7** [siderolabs/talos#8258] [Q&A]
I'm currently building a Talos cluster, 3 nodes, as a home server. As I need persistent storage for things like Nextcloud, I want to use Mayastor, as recommended in the Talos docs.
Installation worked and Mayastor is running. But I have not been able to find a way reserving dedicated cores, as the Mayastor docs suggest.
How can I do this? Is it possible to run Mayastor on specific cores? My idea is, using only e-cores to eventually reduce power consumption.

**Comment by smira** [comment]
Can you point us towards the thing which you couldn't apply to Talos Linux?

**Comment by matthiasarndt7** [comment]
If I understood correctly in its docs, Mayastor would prefer to have two isolated cores to work without conflicts with other workload. I don't see a way to isolate and assign those cores in Talos.

My motivation: After getting Mayastor running, I discovered, it fully utilizes two cores, probably the p-cores, and drives them into maximum turbo frequency. This results in a high power consumption and I don't want that for a permanently running system. So I would like to try assigning two e-cores to eventually reduce consumption significantly.

As a workaround, I currently have the turbo feature of the CPU disabled, keeping frequencies low and consumption in a reasonable range.

**Reply by smira** [reply]
You can use kubelet's feature to isolate CPU cores, I just don't see what is missing. 

I asked for a specific link in the docs on what doesn't apply.

https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/

**Comment by matthiasarndt7** [comment]
Thanks for the new lead! I will give it a try and report back.

The exclusive cores are mentioned in the Mayastor docs, right in the General section:
https://mayastor.gitbook.io/introduction/quickstart/prerequisites
Further in the docs, I found a suggestion, on how to achieve the CPU isolation:
https://mayastor.gitbook.io/introduction/additional-information/performance-tips
But for my understanding, the described way is not valid for Talos as an immutable system, is it? So I decided to ask here.

**Reply by smira** [reply]
It seems to talk about extra kernel args, that is available via `machine.install.extraKernelArgs` followed by an upgrade of Talos.

But not sure how it plays with kubelet CPU manager. 

Either way, Talos Linux doesn't block from using any of that

**Comment by ryayon** [comment]
Hello @matthiasarndt7 ,

I have been able to force `io-engine` using only efficient cores.
My k8s cluster is based on VMs (KVM).
I pinned, for each VMs, the 2 first vCPUs as efficient cores.
Just follow this [libvirt documentation](https://libvirt.org/formatdomain.html#cpu-tuning) and specifically the following block:
```
<domain>
  ...
  <cputune>
    <vcpupin vcpu="0" cpuset="1-4,^2"/>
    <vcpupin vcpu="1" cpuset="0,1"/>
    <vcpupin vcpu="2" cpuset="2,3"/>
    <vcpupin vcpu="3" cpuset="0,4"/>
  </cputune>
```
Once done, I just had to follow the [Mayastor documentation](https://mayastor.gitbook.io/introduction/additional-information/performance-tips#update-mayastor-helm-chart-for-cpu-core-specification) by using `io_engine.coreList={0}`.
As a reminder, `0` because it is the first core which I pinned as efficient core in KVM.
In addition, I didn't need two CPUs as it is for a Homelab.

I hope it will help you and others.

Anyway, it is not related to Talos Linux as mentionned above.

By the way, thanks to all contributors :-)

**Reply by matthiasarndt7** [reply]
Thanks for sharing your approach.

I run Talos on bare metal which does not allow me to tailor the nodes like that, unfortunately.

Anyway, I tried your idea of limiting Mayastor to only one CPU and did some performance tests with the fio pod from the Mayastor docs. I run my E-Cores at 900 MHz and I have a gigabit network in the cluster. Non-fancy SATA SSDs are used.

1 Core, 1 Replica: 57,1 MB/s
1 Core, 3 Replicas: 28,0 MB/s
2 Cores, 1 Replica: 125,0 MB/s
2 Cores, 3 Replica: 55,4 MB/s

I think in the last case I hit the network limit. The node running the pod sends data to the two other nodes over one gigabit link. And even with 4 GHz P-Cores I did not see better performance.

As I found power savings on reducing the core count is really small, I stayed with two.

**Reply by ryayon** [reply]
Thank you for your tests.
Good to know.

**Comment by matthiasarndt7** [comment]
I'm finally done experimenting. Here are my results:

With the Kubelet approach, I was not able to push the Mayastor pods to a specific set of CPUs. I tried to reserve CPU 0-3 (P-Cores) for Kubelet and system but no change was noticeable. I could not find a way to check the per core usage to be really sure. So, my indicator was power consumption, which did not change. This makes me believe, Mayastor still used the P-Cores somehow.

But what actually works is the extraKernelArgs approach together the core assignment as described in the Mayastor Docs. Thanks for putting me on the right path! 

Here is the MachineConfig patch:
```
machine:
  install:
    extraKernelArgs:
      - isolcpus=6,7
      - cpufreq.default_governor=powersave    # optional, I wanted to save power but could not notice any effect
```

To assign cores to Mayastor I used the following command. The command in the docs did not work for me.
```
helm upgrade mayastor mayastor/mayastor -n mayastor --reuse-values --set 'io_engine.coreList={6,7}'
```

Now I could clearly see a significant drop in power consumption, as the E-Cores are utilized.

If you want to do this on a Secure Boot system, as I do, you need to inject the kernel arguments directly in the Talos image using the Image Factory. The kernel arguments in MachineConfig are ignored in the Secure Boot environment.
https://factory.talos.dev/
https://www.talos.dev/v1.6/learn-more/image-factory/
Just put the kernel arguments in the field like this:
```
isolcpus=6,7 cpufreq.default_governor=powersave
```
